// @virtual-list/vue/dist/v2/index.mjs
import { defineComponent as q, ref as I, computed as f, reactive as Z, watch as ee, onMounted as te, nextTick as O } from "vue-demi";
import * as x from "helper-js";
/*!
 * @virtual-list/vue
 * Author: phphe <phphe@outlook.com> (https://github.com/phphe)
 * Homepage: https://virtual-list.phphe.com
 * Released under the MIT License.
 */
const ne = q({
  props: {
    table: Boolean
  }
});
var re = function() {
  var e = this, i = e.$createElement, a = e._self._c || i;
  return e.table ? a("table", [e._t("prepend"), a("tbody", [e._t("default")], 2), e._t("append")], 2) : a("div", [e._t("prepend"), e._t("default"), e._t("append")], 2);
}, ie = [];
function H(e, i, a, s, u, z, w, C) {
  var o = typeof e == "function" ? e.options : e;
  i && (o.render = i, o.staticRenderFns = a, o._compiled = !0);
  var m;
  if (u && (m = u), m)
    if (o.functional) {
      o._injectStyles = m;
      var b = o.render;
      o.render = function(T, y) {
        return m.call(y), b(T, y);
      };
    } else {
      var h = o.beforeCreate;
      o.beforeCreate = h ? [].concat(h, m) : [m];
    }
  return {
    exports: e,
    options: o
  };
}
const A = {};
var le = /* @__PURE__ */ H(
  ne,
  re,
  ie,
  !1,
  oe
);
function oe(e) {
  for (let i in A)
    this[i] = A[i];
}
const ae = /* @__PURE__ */ function() {
  return le.exports;
}(), se = q({
  components: { VirtualListTable: ae },
  props: {
    items: Array,
    disabled: Boolean,
    horizontal: Boolean,
    firstRender: { type: Number, default: 10 },
    buffer: { type: Number, default: 100 },
    itemKey: {
      type: [String, Function]
    },
    itemSize: {
      type: Function
    },
    table: Boolean
  },
  setup(e) {
    const i = I(0), a = I(e.firstRender - 1), s = f(
      () => {
        var t;
        return x.notGreaterThan(a.value, (((t = e.items) == null ? void 0 : t.length) || 1) - 1);
      }
    ), u = I(0), z = f(() => d.value[i.value] ? k(i.value) : 0), w = f(
      () => d.value.length > 0 ? k(d.value.length - 1) + x.arrayLast(h.value) : 0
    ), C = f(
      () => d.value[s.value] ? w.value - k(s.value) - h.value[s.value] : 0
    ), o = f(
      () => e.disabled ? {} : { overflow: "auto" }
    ), m = f(() => {
      const t = {
        display: "flex"
      };
      return e.disabled || (e.horizontal ? Object.assign(t, {
        "margin-left": z.value + "px",
        "margin-right": C.value + "px",
        width: w.value - C.value - z.value + "px"
      }) : Object.assign(t, {
        "margin-top": z.value + "px",
        "margin-bottom": C.value + "px"
      })), t["flex-direction"] = e.horizontal ? "row" : "column", e.table && (delete t.display, delete t["flex-direction"]), t;
    }), b = f(
      () => Z((e.items || []).map(() => null))
    ), h = f(
      () => (e.items || []).map((t, n) => {
        var _;
        if (b[n] != null)
          return b[n];
        let r = (_ = e.itemSize) == null ? void 0 : _.call(e, t, n);
        return r == null && (r = u.value), r;
      })
    ), d = f(() => {
      const t = [];
      return h.value.reduce((n, r) => (t.push(n), n + r), 0), t;
    });
    ee(() => e.items, g);
    const T = f(() => {
      if (!e.items || e.disabled)
        return;
      const t = [];
      for (let n = i.value; n <= s.value; n++) {
        const r = e.items[n];
        if (!r)
          break;
        t.push({ item: r, index: n });
      }
      return t;
    }), y = I(), B = I();
    te(async () => {
      g();
      try {
        D();
      } catch {
        await O(), g();
      }
    });
    let L;
    function G() {
      const t = y.value;
      if (!t)
        return;
      const n = $(t);
      L != null && e.buffer - Math.abs(n - L) >= 10 || (L = n, g());
    }
    let N = !1, W = !1;
    async function g() {
      var P;
      if (N) {
        W = !0;
        return;
      }
      if (!e.items || e.disabled)
        return;
      N = !0;
      const t = y.value, n = (P = B.value) == null ? void 0 : P.$el;
      if (!t || !n)
        return;
      u.value || (u.value = X()), i.value = Q(), a.value = U(), await O();
      let r, _ = 0;
      const F = {}, M = e.table ? n.querySelector("tbody").children : n.children;
      for (let c = 0; c < M.length; c++) {
        const l = M[c], v = x.css(l, "position");
        if (v && ["absolute", "fixed"].includes(v))
          continue;
        const S = x.css(l, "display") !== "none" ? j(l) : 0, p = l.getAttribute("vt-index"), R = p ? parseInt(p) : i.value + _;
        F[R] = (F[R] || 0) + S, _++;
      }
      for (const c of Object.keys(F)) {
        const l = parseInt(c);
        b.value[l] !== F[l] && (b.value[l] = F[l], r = !0);
      }
      r && await O(), N = !1, W && (W = !1, g());
      function Q() {
        const c = $(t) - E(t) - e.buffer;
        return x.binarySearch(
          d.value,
          (v) => v - c,
          { returnNearestIfNoHit: !0 }
        ).index;
      }
      function U() {
        const c = $(t) - E(t) + K(t) + e.buffer;
        return x.binarySearch(
          d.value,
          (v) => v - c,
          { returnNearestIfNoHit: !0 }
        ).index;
      }
      function X() {
        const l = [], v = e.table ? n.querySelector("tbody").children : n.children;
        for (let S = 0; S < v.length; S++) {
          const p = v[S], R = getComputedStyle(p);
          if (["absolute", "fixed"].includes(R.position))
            continue;
          const Y = j(p);
          if (l.push(Y), l.length >= 10)
            break;
        }
        return l.length === 0 ? 0 : l.reduce((S, p) => S + p, 0) / l.length;
      }
    }
    function K(t) {
      const n = getComputedStyle(t);
      let r = parseFloat(e.horizontal ? n.width : n.height);
      return n.boxSizing === "border-box" && (e.horizontal ? r = r - parseFloat(n.borderLeftWidth) - parseFloat(n.borderRightWidth) : r = r - parseFloat(n.borderTopWidth) - parseFloat(n.borderBottomWidth)), r;
    }
    function j(t) {
      let n = K(t);
      const r = getComputedStyle(t);
      return e.horizontal ? n += parseFloat(r.borderLeftWidth) + parseFloat(r.borderRightWidth) + parseFloat(r.marginLeft) + parseFloat(r.marginRight) : n += parseFloat(r.borderTopWidth) + parseFloat(r.borderBottomWidth) + parseFloat(r.marginTop) + parseFloat(r.marginBottom), n = Number.isNaN(n) ? 0 : n, n;
    }
    function $(t) {
      return e.horizontal ? t.scrollLeft : t.scrollTop;
    }
    function E(t) {
      const n = getComputedStyle(t);
      return e.horizontal ? parseFloat(n.paddingLeft) : parseFloat(n.paddingTop);
    }
    function k(t) {
      return d.value[t];
    }
    function D() {
      const t = y.value;
      new ResizeObserver((r) => {
        for (let _ of r)
          if (x.hasClass(_.target, "vtlist")) {
            g();
            break;
          }
      }).observe(t);
    }
    function J(t, n) {
      if (e.itemKey) {
        if (typeof e.itemKey == "string" && e.itemKey === "index")
          return n;
        if (typeof e.itemKey == "function")
          return e.itemKey(t, n);
      }
    }
    return {
      listElRef: y,
      listInnerRef: B,
      onscroll: G,
      listStyle: o,
      listInnerStyle: m,
      visibleItemsInfo: T,
      getItemKey: J,
      update: g,
      sizes: h,
      positions: d,
      runtimeSizes: b
    };
  }
});
var ue = function() {
  var e = this, i = e.$createElement, a = e._self._c || i;
  return a("div", { ref: "listElRef", staticClass: "vtlist", style: e.listStyle, on: { "&scroll": function(s) {
    return e.onscroll.apply(null, arguments);
  } } }, [a("VirtualListTable", { ref: "listInnerRef", staticClass: "vtlist-inner", style: e.listInnerStyle, attrs: { table: e.table }, scopedSlots: e._u([{ key: "prepend", fn: function() {
    return [e._t("prepend")];
  }, proxy: !0 }, { key: "append", fn: function() {
    return [e._t("append")];
  }, proxy: !0 }], null, !0) }, [e.disabled ? [e._l(e.items, function(s, u) {
    return [e._t("default", null, { item: s, index: u })];
  })] : [e._l(e.visibleItemsInfo, function(s) {
    var u = s.item, z = s.index;
    return [e._t("default", null, { item: u, index: z })];
  })]], 2)], 1);
}, ce = [];
const V = {};
var fe = /* @__PURE__ */ H(
  se,
  ue,
  ce,
  !1,
  de
);
function de(e) {
  for (let i in V)
    this[i] = V[i];
}
const me = /* @__PURE__ */ function() {
  return fe.exports;
}();
export {
  me as default
};
